Mathlib/GroupTheory/Coxeter/Inversion.lean

```diff
@@ -0,0 +1,470 @@
+  constructor
+  · exact cs.isReflection_of_mem_rightInvSeq ω ht
+  · obtain ⟨⟨j, hj⟩, rfl⟩ := List.mem_iff_get.mp ht
+    rw [← List.getD_eq_get _ 1 hj, wordProd_mul_getD_rightInvSeq]
+    rw [cs.length_rightInvSeq] at hj
+    calc
+      ℓ (π (ω.eraseIdx j))
+      _ ≤ (ω.eraseIdx j).length   := cs.length_wordProd_le _
+      _ < ω.length                := by rw [← List.length_eraseIdx_add_one hj]; exact lt_add_one _
+      _ = ℓ (π ω)                 := hω.symm
+
+theorem isLeftInversion_of_mem_leftInvSeq {ω : List B} (hω : cs.IsReduced ω) {t : W}
+    (ht : t ∈ lis ω) : cs.IsLeftInversion (π ω) t := by
+  constructor
+  · exact cs.isReflection_of_mem_leftInvSeq ω ht
+  · obtain ⟨⟨j, hj⟩, rfl⟩ := List.mem_iff_get.mp ht
+    rw [← List.getD_eq_get _ 1 hj, getD_leftInvSeq_mul_wordProd]
+    rw [cs.length_leftInvSeq] at hj
+    calc
+      ℓ (π (ω.eraseIdx j))
+      _ ≤ (ω.eraseIdx j).length   := cs.length_wordProd_le _
+      _ < ω.length                := by rw [← List.length_eraseIdx_add_one hj]; exact lt_add_one _
+      _ = ℓ (π ω)                 := hω.symm
+
+theorem prod_rightInvSeq (ω : List B) : prod (ris ω) = (π ω)⁻¹ := by
+  induction' ω with i ω ih
+  · simp
+  · simp [rightInvSeq, ih, wordProd_cons]
+
+theorem prod_leftInvSeq (ω : List B) : prod (lis ω) = (π ω)⁻¹ := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse, prod_reverse_noncomm]
+  have : List.map (fun x ↦ x⁻¹) (ris ω.reverse) = ris ω.reverse := calc
+    List.map (fun x ↦ x⁻¹) (ris ω.reverse)
+    _ = List.map id (ris ω.reverse)             := by
+        apply List.map_congr
+        intro t ht
+        exact (cs.isReflection_of_mem_rightInvSeq _ ht).inv
+    _ = ris ω.reverse                           := map_id _
+  rw [this]
+  nth_rw 2 [← reverse_reverse ω]
+  rw [wordProd_reverse]
+  exact cs.prod_rightInvSeq _
+
+theorem nodup_rightInvSeq_of_reduced {ω : List B} (rω : cs.IsReduced ω) : List.Nodup (ris ω) := by
+  apply List.nodup_iff_get?_ne_get?.mpr
+  intro j j' j_lt_j' j'_lt_length dup
+  -- dup : get? (rightInvSeq cs ω) j = get? (rightInvSeq cs ω) j'
+  -- ⊢ False
+  simp at j'_lt_length
+  -- j'_lt_length: j' < List.length ω
+  rw [get?_eq_get (by simp; linarith), get?_eq_get (by simp; linarith)] at dup
+  apply Option.some_injective at dup
+  rw [← getD_eq_get _ 1, ← getD_eq_get _ 1] at dup
+  set! t := (ris ω).getD j 1 with h₁
+  set! t' := (ris (ω.eraseIdx j)).getD (j' - 1) 1 with h₂
+  have h₃ : t' = (ris ω).getD j' 1                    := by
+    rw [h₂]
+    rw [cs.getD_rightInvSeq, cs.getD_rightInvSeq]
+    rw [(Nat.sub_add_cancel (by linarith) : j' - 1 + 1 = j')]
+    rw [eraseIdx_eq_take_drop_succ]
+    rw [drop_append_eq_append_drop]
+    rw [drop_length_le (by simp; left; linarith)]
+    rw [length_take, drop_drop, nil_append]
+    rw [min_eq_left_of_lt (j_lt_j'.trans j'_lt_length)]
+    rw [Nat.succ_eq_add_one, ← add_assoc, Nat.sub_add_cancel (by linarith)]
+    rw [mul_left_inj, mul_right_inj]
```

Comment:
I think you can chain these together in a single `rw`, even if you format them on separate lines.
If you want to keep the line breaks, you should probably add a small comment, otherwise someone else will merge the lines in the future.